lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -386,16 +435,48 @@ prefilterBlock b u0 = runState $ do
         => ([(Tx, TxMeta)], UTxO)
         -> Tx
         -> State s ([(Tx, TxMeta)], UTxO)
-    applyTx (!txs, !u) tx = do
-        ourU <- state $ utxoOurs tx
-        let ourIns = Set.fromList (inputs tx) `Set.intersection` dom (u <> ourU)
-        let u' = (u <> ourU) `excluding` ourIns
+    applyTx (!txs, !prevUTxO) tx = do
+        -- The next UTxO state (apply a state transition) (e.g. remove
+        -- transaction outputs we've spent).
+        ourNextUTxO <-
+            (spendTx tx prevUTxO <>) <$> filterByAddressM isOurs' (utxoFromTx tx)
+
         ourWithdrawals <- Coin . sum . fmap (unCoin . snd) <$>
             mapMaybeM ourWithdrawal (Map.toList $ withdrawals tx)
-        let received = balance ourU
-        let spent = balance (u `restrictedBy` ourIns) `TB.add` TB.fromCoin ourWithdrawals
-        let hasKnownInput = ourIns /= mempty
-        let hasKnownOutput = ourU /= mempty
+
+        let received = balance (ourNextUTxO `excluding` dom prevUTxO)
+        let spent =
+                balance (prevUTxO `excluding` dom ourNextUTxO)
+                `TB.add` TB.fromCoin ourWithdrawals
+
+        (ownedAndKnownTxIns, ownedAndKnownTxOuts) <- do
+            -- A new transaction expands the set of transaction inputs/outputs
+            -- we know about:
+            let known = prevUTxO <> utxoFromTx tx
+            -- But not all those transaction inputs/outputs belong to us:
+            ownedAndKnown <- filterByAddressM isOurs' known
```

Comment:
Hi Sam. I suspect that evaluating this expression will still re-filter the wallet's whole UTxO, since:
```hs
   filterByAddressM isOurs' known
== filterByAddressM isOurs' (prevUTxO <> utxoFromTx tx)
```
I'm wondering, is there any way we can avoid refiltering parts of the UTxO set that have been filtered before?

Perhaps:
```suggestion
            ownedAndKnown <- (prevUTxO <>)
                <$> filterByAddressM isOurs' (utxoFromTx tx)
```

(Apologies if I've misunderstood something!)

Reply:
True! Yep, we can just do what you suggested, I will fix this locally.